utils/aplay/alsa-pcm.c

```diff
@@ -236,23 +261,23 @@ int alsa_pcm_write(
 	snd_pcm_sframes_t written_frames = 0;
 	snd_pcm_uframes_t hw_avail = pcm->buffer_frames - avail;
 
-	/* If not draining, write only as many frames as possible without
-	 * blocking. If necessary insert silence frames to prevent underrun. */
 	if (!drain) {
-		if (frames > avail)
-			frames = avail;
-		else if (hw_avail + frames < pcm->underrun_threshold &&
+		if (frames == 0 && hw_avail < pcm->period_frames &&
 					snd_pcm_state(pcm->pcm) == SND_PCM_STATE_RUNNING) {
-			/* Pad the buffer with enough silence to restore it to the underrun
-			 * threshold. */
-			const size_t padding_frames = pcm->underrun_threshold - (hw_avail + frames);
-			const size_t padding_samples = padding_frames * pcm->channels;
-			if (verbose >= 3)
-				info("Underrun imminent: inserting %zu silence frames", padding_frames);
-			snd_pcm_format_set_silence(pcm->format, buffer->tail, padding_samples);
-			ffb_seek(buffer, padding_samples);
-			frames += padding_frames;
+			/* When the stream runs dry we drain the ALSA buffer and leave the
+			 * ALSA device stopped until fresh frames arrive from the server. */
+			debug("draining ALSA playback PCM to avoid underrun");
+			snd_pcm_drain(pcm->pcm);
+			snd_pcm_prepare(pcm->pcm);
+			pcm->delay = 0;
+			pcm->hw_avail = 0;
+			/* Flag an underrun to indicate that there has been a discontinuity
+			 * in the input stream. */
+			pcm->underrun = true;
```

Comment:
Isn't `return 0;` missing here?

I've applied this commit on current master and I think I will push it to github. After some tests I'm not sure whether this approach is better or not, but it is definitely simpler (so if the effect is the same, simplicity wins). Also, during tests I'm seeing mostly debug log `ALSA playback PCM underrun` which is from the `snd_pcm_avail_delay()` check. So, it seems that the code responsible for inserting silent frames was rarely used anyway...

Reply:
> Isn't return 0; missing here?

Yes ! sorry :disappointed:

Reply:
Nothing to be sorry, that's what for the review process is :D

Also, please check GitHub invitations. I'm not sure whether it will be of any use for you, but treat it like a gift for personal use.

Reply:
I'm now reducing my time spent on software development (due to health issues, I am an old man now). I will continue to contribute comments and documentation to bluez-alsa for as long as I am using it myself, but this PR and my multi-client branch will most likely be my last original code contributions (I also have some plugin code for ALSA swparams support, but there do not appear to be any applications that use stop_threshold, silence_threshold, etc so there is little benefit in maintaining it in master). So, thank you for your confidence in me to invite me to a new project, but unfortunately I have to decline.

Reply:
> So, thank you for your confidence in me to invite me to a new project, but unfortunately I have to decline.

It's not a project which needs any development I think. You can use it with bluez-alsa as it is. But of course it's up to you.

Also, once again many thanks for your excellent contributions!